src/globus_sdk/transport/_clientinfo.py

```diff
@@ -0,0 +1,115 @@
+
+_RESERVED_CHARS = ";,="
+
+
+class GlobusClientInfo:
+    """
+    An implementation of X-Globus-Client-Info as an object. This header encodes a
+    mapping of multiple products to versions and potentially other information.
+
+    Values can be added to a clientinfo object via the ``add()`` method.
+
+    .. rubric:: ``X-Globus-Client-Info`` Specification
+
+    Header Name: ``X-Globus-Client-Info``
+
+    Header Value:
+
+    - A semicolon (``;``) separated list of client information.
+
+    - Client information is a comma-separated list of ``=`` delimited key-value pairs.
+      Well-known values for client-information are:
+
+      - ``product``: A unique identifier of the product.
+      - ``version``: Relevant version information for the product.
+
+    - Based on the above, the characters ``;,=`` should be considered reserved and
+      should NOT be included in client information values to ensure proper parsing.
+
+    .. rubric:: Example Headers
+
+    .. code-block:: none
+
+        X-Globus-ClientInfo: product=python-sdk,version=3.32.1
+        X-Globus-ClientInfo: product=python-sdk,version=3.32.1;product=cli,version=4.0.0a1
+
+    .. note::
+
+        The ``GlobusClientInfo`` object is not guaranteed to reject all invalid usages.
+        For example, ``product`` is required to be unique per header, and users are
+        expected to enforce this in their usage.
+    """  # noqa: E501
+
+    def __init__(self) -> None:
+        self.infos: list[str] = []
+
+    def __bool__(self) -> bool:
+        """Check if there are any values present."""
+        return bool(self.infos)
+
+    def format(self) -> str:
+        """Format as a header value."""
+        return ";".join(self.infos)
+
+    def add(self, value: str | dict[str, str]) -> None:
+        """
+        Add an item to the clientinfo. The item is either already formatted
+        as a string, or is a dict containing values to format.
+
+        :param value: The element to add to the client-info. If it is a dict,
+            it may not contain reserved characters in any keys or values. If it is a
+            string, it cannot contain the ``;`` separator.
+        """
+        if not isinstance(value, str):
+            value = ",".join(_format_items(value))
+        elif ";" in value:
+            raise exc.GlobusSDKUsageError(
+                "GlobusClientInfo.add() cannot be used to add multiple items in "
+                "an already-joined string. Add items separately instead. "
+                f"Bad usage: '{value}'"
+            )
+        self.infos.append(value)
+
+
+class DefaultGlobusClientInfo(GlobusClientInfo):
+    """
+    A variant of GlobusClientInfo which always initializes itself to start with
+
+        product=python-sdk,version=...
+
+    using the current package version information.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.add({"product": "python-sdk", "version": __version__})
```

Comment:
It seems like this information should _always_ be tacked on at serialization time.

That is to say: devs who are using the SDK will always have critical SDK client information present.

Therefore I think that there should be no `DefaultGlobusClientInfo` class, but instead the SDK client information should be hard-coded to serialize into the HTTP header (and, probably, be the _first_ client info in the serialized header, in case the remainder of the header is b0rked for any reason and cannot be deserialized).

Reply:
Okay, let's merge it into the superclass so that it's part of the base implementation here. I buy a lot of the thinking here.

The place where I want to exercise some caution is around baking it in so deeply that removing, inspecting, or manipulating it requires subclassing the transport class itself. I'd therefore prefer to keep it as structured data, but structured data which is populated by default.